upi/src/main/java/com/adyen/checkout/upi/internal/ui/model/UPIIntentItem.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2024 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by ararat on 22/4/2024.
+ */
+
+package com.adyen.checkout.upi.internal.ui.model
+
+import androidx.annotation.StringRes
+import com.adyen.checkout.core.Environment
+
+internal sealed class UPIIntentItem {
+
+    abstract fun areItemsTheSame(newItem: UPIIntentItem): Boolean
+    abstract fun areContentsTheSame(newItem: UPIIntentItem): Boolean
+    abstract fun getChangePayload(newItem: UPIIntentItem): Any?
+
+    data class PaymentApp(
+        val id: String,
+        val name: String,
+        val environment: Environment,
+    ) : UPIIntentItem() {
+        override fun areItemsTheSame(newItem: UPIIntentItem) =
+            newItem is PaymentApp &&
+                id == newItem.id
+
+        override fun areContentsTheSame(newItem: UPIIntentItem) =
+            newItem is PaymentApp &&
+                name == newItem.name &&
+                environment == newItem.environment
+
+        override fun getChangePayload(newItem: UPIIntentItem) = null
+    }
+
+    data object GenericApp : UPIIntentItem() {
+        override fun areItemsTheSame(newItem: UPIIntentItem) = newItem is GenericApp
+        override fun areContentsTheSame(newItem: UPIIntentItem) = newItem is GenericApp
+        override fun getChangePayload(newItem: UPIIntentItem) = null
+    }
+
+    data class ManualInput(
+        @StringRes val errorMessageResource: Int?
+    ) : UPIIntentItem() {
+        override fun areItemsTheSame(newItem: UPIIntentItem) = newItem is ManualInput
+        override fun areContentsTheSame(newItem: UPIIntentItem) =
+            newItem is ManualInput &&
+                errorMessageResource == newItem.errorMessageResource
+
+        // This has been implemented to avoid creating a new ViewHolder when the input field has validation error
+        override fun getChangePayload(newItem: UPIIntentItem) =
```

Comment:
This is probably caused by the error message resource changing. I think this could be solved by also passing the entered text as state to the view holder.

Also, while testing I run into this weird error:
```
java.lang.IllegalArgumentException: parameter must be a descendant of this view
	at android.view.ViewGroup.offsetRectBetweenParentAndChild(ViewGroup.java:6323)
	at android.view.ViewGroup.offsetDescendantRectToMyCoords(ViewGroup.java:6252)
	at androidx.core.widget.NestedScrollView.isWithinDeltaOfScreen(NestedScrollView.java:1735)
	at androidx.core.widget.NestedScrollView.onSizeChanged(NestedScrollView.java:2284)
```
It happens consistently when I triggered a validation error and then select another UPI app. I did remove `getChangePayload`, but I think that was masking this error. Apparently, it has something to do with edit texts in recyclerviews 🤔

Reply:
I did some investigation for the crash and it is happening only when bottom sheet changes its size. If the bottom sheet opens in full screen, because of the keyboard visibility, the crash does not happen. 
The crash is related to the handling of `focus` and `unfocus` states of the `editTextManualAddress`. When `bindItem()` is called in `UPIIntentManualAddressViewHolder`, we request/clear the focus of the `editTextManualAddress`. What we also do in `bindItem()` is to hide the `textInputLayoutManualAddress`, which is the parent of `editTextManualAddress`. That way when selecting another UPI option, we hide the extra views which show only when selecting the manual input option.
So it crashes when we call `clearFocus()` for `editTextManualAddress` when the item is deselected (because we clicked on another item) and at the same time we hide the `textInputLayoutManualAddress`. When we hide the `textInputLayoutManualAddress`, the bottom sheet changes its size and it happens before the `editTextManualAddress` cleared its focus. Strangely something goes wrong at that part in the view logic in combination with multiple view holders, which causes the `java.lang.IllegalArgumentException: parameter must be a descendant of this view` internal exception.

The solution that I had was to override the `getChangePayload()`, which forces the recycler view to update the existing view holder with the payload. Since there is a payload to be passed to the view holder, it does not create a new view holder.

If we do not override the `getChangePayload()` function, when the ManualInput changes, the recycler view will try to create a view holder, which in turn will hide and show input fields with previous values, causing a UI glitch like in [here](https://github.com/Adyen/adyen-android/pull/1615#discussion_r1634767878).

I tested the UPI integration with overriding the `getChangePayload()` and I am not able to reproduce the crash. If you also agree, we can go with this solution, ask the team to help us extensively test the UPI component and later improve it if the UI changes.